// winston-azure-blob-transport/lib/blobTransport.js
(function() {
  var BlobTransport, MAX_BLOCK_SIZE, Promise, Transport, _, async, azure, chunk, debug, errorToJson, util, winston,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    extend = function(child, parent) { for (var key in parent) { if (hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; },
    hasProp = {}.hasOwnProperty;

  debug = require("debug")("winston-blob-transport");

  _ = require("lodash");

  util = require("util");

  errorToJson = require("error-to-json");

  azure = require("azure-storage");

  async = require("async");

  winston = require("winston");

  chunk = require("chunk");

  Promise = require("bluebird");

  Transport = winston.Transport;

  MAX_BLOCK_SIZE = azure.Constants.BlobConstants.MAX_BLOCK_SIZE;

  BlobTransport = (function(superClass) {
    extend(BlobTransport, superClass);

    function BlobTransport(arg) {
      var ref;
      this.account = arg.account, this.containerName = arg.containerName, this.blobName = arg.blobName, this.level = (ref = arg.level) != null ? ref : "info";
      this._buildClient = bind(this._buildClient, this);
      this._meta = bind(this._meta, this);
      this._formatLine = bind(this._formatLine, this);
      this._retryIfNecessary = bind(this._retryIfNecessary, this);
      this._buildCargo = bind(this._buildCargo, this);
      this.log = bind(this.log, this);
      this.name = "BlobTransport";
      this.cargo = this._buildCargo();
      this.client = this._buildClient(this.account);
    }

    BlobTransport.prototype.initialize = function() {
      return Promise.promisifyAll(azure.createBlobService(this.account.name, this.account.key)).createContainerIfNotExistsAsync(this.containerName, {
        publicAccessLevel: "blob"
      }).then((function(_this) {
        return function(created) {
          return debug("Container: " + _this.container + " - " + (created ? 'creada' : 'existente'));
        };
      })(this));
    };

    BlobTransport.prototype.log = function(level, msg, meta, callback) {
      var line;
      line = this._formatLine({
        level: level,
        msg: msg,
        meta: meta
      });
      this.cargo.push({
        line: line,
        callback: callback
      });
    };

    BlobTransport.prototype._buildCargo = function() {
      return async.cargo((function(_this) {
        return function(tasks, __whenFinishCargo) {
          var __whenLogAllBlock, chunks, logBlock;
          __whenLogAllBlock = function() {
            debug("Finish append all lines to blob");
            _.each(tasks, function(arg) {
              var callback;
              callback = arg.callback;
              return callback(null, true);
            });
            return __whenFinishCargo();
          };
          debug("Log " + tasks.length + "th lines");
          logBlock = _.map(tasks, "line").join("");
          debug("Starting append log lines to blob. Size " + logBlock.length);
          chunks = chunk(logBlock, MAX_BLOCK_SIZE);
          debug("Saving " + chunks.length + " chunk(s)");
          return async.eachSeries(chunks, function(chunk, whenLoggedChunk) {
            debug("Saving log with size " + chunk.length);
            return _this.client.appendFromText(_this.containerName, _this.blobName, chunk, function(err, result) {
              if (err) {
                return _this._retryIfNecessary(err, chunk, whenLoggedChunk);
              }
              return whenLoggedChunk();
            });
          }, function(err) {
            if (err) {
              debug("Error in block");
            }
            return __whenLogAllBlock();
          });
        };
      })(this));
    };

    BlobTransport.prototype._retryIfNecessary = function(err, block, whenLoggedChunk) {
      var __createAndAppend, __doesNotExistFile, __handle;
      __createAndAppend = (function(_this) {
        return function() {
          return _this.client.createAppendBlobFromText(_this.containerName, _this.blobName, block, {}, __handle);
        };
      })(this);
      __doesNotExistFile = function() {
        return (err.code != null) && err.code === "NotFound";
      };
      __handle = function(err) {
        if (err) {
          debug("Error in append", err);
        }
        return whenLoggedChunk();
      };
      if (__doesNotExistFile()) {
        return __createAndAppend();
      } else {
        return __handle(err);
      }
    };

    BlobTransport.prototype._formatLine = function(arg) {
      var level, meta, msg;
      level = arg.level, msg = arg.msg, meta = arg.meta;
      return "[" + level + "] - " + (this._timestamp()) + " - " + msg + " " + (this._meta(meta)) + " \n";
    };

    BlobTransport.prototype._timestamp = function() {
      return new Date().toISOString();
    };

    BlobTransport.prototype._meta = function(meta) {
      if (meta instanceof Error) {
        meta = errorToJson(meta);
      }
      if (_.isEmpty(meta)) {
        return "";
      } else {
        return "- " + (util.inspect(meta));
      }
    };

    BlobTransport.prototype._buildClient = function(arg) {
      var key, name;
      name = arg.name, key = arg.key;
      return azure.createBlobService(name, key);
    };

    return BlobTransport;

  })(Transport);

  winston.transports.AzureBlob = BlobTransport;

  module.exports = BlobTransport;

}).call(this);
