// nodedata/config.js
"use strict";
class Config {
}
Config.DbConnection = "mongodb://localhost:27017/test_nodedata";
Config.basePath = "data";
Config.apiversion = "v1";
Config.ElasticSearchConnection = "http://localhost:9200";
Config.ApplyElasticSearch = false;
Config.ignorePaths = [];
Config.isMultiThreaded = true; // This param is for configuring multi process using worker attribute..
exports.Config = Config;
class SqlConfig {
}
SqlConfig.isSqlEnabled = false;
SqlConfig.database = "test";
SqlConfig.username = "sa";
SqlConfig.password = "Apr@2016";
SqlConfig.sequlizeSetting = {
    host: '172.19.101.120',
    dialect: 'mssql',
    pool: {
        max: 5,
        min: 0,
        idle: 10000
    }
};
exports.SqlConfig = SqlConfig;
class Security {
}
Security.isAutheticationEnabled = "disabled"; //allowed values: "disabled","enabledWithoutAuthorization","enabledWithAuthorization"
Security.authenticationType = "TokenBased"; //allowed values: "passwordBased","TokenBased"
Security.useFaceBookAuth = false;
exports.Security = Security;
class facebookAuth {
}
facebookAuth.clientID = '11'; // your App ID
facebookAuth.clientSecret = 'aa'; // your App Secret
facebookAuth.callbackURL = 'http://localhost:23548/auth/facebook/callback';
exports.facebookAuth = facebookAuth;
