// nodedata/mongoose/utils/index.js
"use strict";
const Mongoose = require('mongoose');
const entity_change_1 = require('../../core/enums/entity-change');
const Enumerable = require('linq');
const db_1 = require('../db');
const model_entity_1 = require('../../core/dynamic/model-entity');
const decorators_1 = require('../../core/constants/decorators');
const mongoose_1 = require("mongoose");
function castToMongooseType(value, schemaType) {
    var newVal;
    switch (schemaType) {
        case Mongoose.Types.ObjectId:
            if (value instanceof Mongoose.Types.ObjectId) {
                newVal = value;
            }
            else if (typeof value === 'string') {
                newVal = new Mongoose.Types.ObjectId(value);
            }
            else {
                throw 'cannot cast to primary key type';
            }
            break;
        case String:
            if (typeof value === 'string') {
                newVal = value;
            }
            newVal = value.toString();
            break;
        case Number:
            if (typeof value === 'number') {
                newVal = value;
            }
            newVal = parseInt(value);
            if (isNaN(newVal)) {
                throw 'cannot cast to primary key type';
            }
            break;
        default:
            newVal = value;
            break;
    }
    return newVal;
}
exports.castToMongooseType = castToMongooseType;
function getPropertiesFromObject(changedObj) {
    return Enumerable.from(changedObj).select((x) => x.key).toArray();
}
exports.getPropertiesFromObject = getPropertiesFromObject;
/**
 * return json from mongoose object
 * @param result
 */
function toObject(result) {
    if (result instanceof Array) {
        return Enumerable.from(result).select((x) => x.toObject()).toArray();
    }
    return result ? result.toObject() : null;
}
exports.toObject = toObject;
/**
 * It creates list of properties to set/unset/push.
 * If array is passed, then for put whole array is replaced but for patch array is updated. For e.g.
 * Case 'put': Suppose there is an object {'ids':['1']}. On put {'ids':['2']}, it will result {'ids':['2']}.
 * Case 'patch': Suppose there is an object {'ids':['1']}. On patch {'ids':['2']}, it will result {'ids':['1', '2']}.
 * @param obj
 * @param type
 */
function getUpdatedProps(obj, type) {
    var push = {};
    var set = {};
    var unset = {};
    var s = false, u = false, p = false;
    for (var i in obj) {
        if (obj[i] == undefined || obj[i] == null || obj[i] == undefined && obj[i] == '' || (obj[i] instanceof Array && obj[i] == []) || obj[i] == {}) {
            unset[i] = obj[i];
            u = true;
        }
        else {
            if (type == entity_change_1.EntityChange.patch && obj[i] instanceof Array) {
                push[i] = {
                    $each: obj[i]
                };
                p = true;
            }
            else {
                set[i] = obj[i];
                s = true;
            }
        }
    }
    var json = {};
    if (s) {
        json['$set'] = set;
    }
    if (u) {
        json['$unset'] = unset;
    }
    if (p) {
        json['$push'] = push;
    }
    return json;
}
exports.getUpdatedProps = getUpdatedProps;
function isPropertyUpdateRequired(changedProps, properties) {
    if (properties && properties.length > 0) {
        if (Enumerable.from(properties).any(x => changedProps.indexOf(x) > -1))
            return true;
    }
    if (!changedProps || changedProps.length == 0)
        return false;
    else if (!properties || properties.length == 0)
        return true;
    else {
        if (Enumerable.from(properties).any(x => changedProps.indexOf(x) > -1))
            return true;
        else
            return false;
    }
}
exports.isPropertyUpdateRequired = isPropertyUpdateRequired;
function getCurrentDBModel(schemaName) {
    var model = model_entity_1.getModel(schemaName);
    return db_1.getDbSpecifcModel(schemaName, model.schema);
}
exports.getCurrentDBModel = getCurrentDBModel;
function getMongoUpdatOperatorForRelation(meta) {
    var operator = "";
    switch (meta.decorator) {
        case decorators_1.Decorators.ONETOMANY:
            operator = '.$';
            break;
        case decorators_1.Decorators.MANYTOMANY:
            operator = '.$';
            break;
        case decorators_1.Decorators.MANYTOONE:
            operator = "";
            break;
        case decorators_1.Decorators.ONETOONE:
            operator = "";
            break;
    }
    return operator;
}
exports.getMongoUpdatOperatorForRelation = getMongoUpdatOperatorForRelation;
function isBasonOrStringType(obj) {
    if (!obj) {
        return undefined;
    }
    return obj instanceof mongoose_1.Types.ObjectId || typeof obj === "string";
}
exports.isBasonOrStringType = isBasonOrStringType;
