// nodedata/mongoose/mongoose-model.js
"use strict";
const Mongoose = require("mongoose");
const Q = require('q');
const entity_change_1 = require('../core/enums/entity-change');
const model_entity_1 = require('../core/dynamic/model-entity');
const Enumerable = require('linq');
const winstonLog_1 = require('../logging/winstonLog');
const mongooseHelper = require('./mongoose-model-helper');
const CoreUtils = require("../core/utils");
const Utils = require('./utils');
const utils_1 = require("../core/metadata/utils");
const decorators_1 = require('../core/constants/decorators');
const dynamic_repository_1 = require('../core/dynamic/dynamic-repository');
/**
 * Iterate through objArr and check if any child object need to be added. If yes, then add those child objects.
 * Bulk create these updated objects.
 * Usage - Post multiple objects parallely
 * @param model
 * @param objArr
 */
function bulkPost(model, objArr) {
    var addChildModel = [];
    // create all cloned models
    var clonedModels = [];
    Enumerable.from(objArr).forEach(obj => {
        var cloneObj = mongooseHelper.removeTransientProperties(model, obj);
        clonedModels.push(cloneObj);
        addChildModel.push(mongooseHelper.addChildModelToParent(model, cloneObj, null));
    });
    return Q.allSettled(addChildModel)
        .then(result => {
        // autogenerate ids of all the objects
        Enumerable.from(clonedModels).forEach(clonedObj => {
            try {
                mongooseHelper.autogenerateIdsForAutoFields(model, clonedObj);
            }
            catch (ex) {
                winstonLog_1.winstonLog.logError(`Error in bulkPost ${ex}`);
                return Q.reject(ex);
            }
        });
        return Q.nbind(model.create, model)(clonedModels).then(result => {
            return Enumerable.from(result).select(x => Utils.toObject(x)).toArray();
        })
            .catch(error => {
            winstonLog_1.winstonLog.logError(`Error in bulkPost ${error}`);
            return Q.reject(error);
        });
    });
}
exports.bulkPost = bulkPost;
/**
 * Iterate through objArr and call put for these
 * Usage - Update multiple object sequentially
 * @param model
 * @param objArr
 */
function bulkPut(model, objArr) {
    var asyncCalls = [];
    var length = objArr.length;
    var ids = objArr.map(x => x._id);
    var bulk = model.collection.initializeUnorderedBulkOp();
    var asyncCalls = [];
    // classic for loop used gives high performanance
    for (var i = 0; i < length; i++) {
        asyncCalls.push(mongooseHelper.addChildModelToParent(model, objArr[i], objArr[i]._id).then(result => {
            var objectId = new Mongoose.Types.ObjectId(result._id);
            delete result._id;
            let clonedObj = mongooseHelper.removeTransientProperties(model, result);
            var updatedProps = Utils.getUpdatedProps(clonedObj, entity_change_1.EntityChange.put);
            bulk.find({ _id: objectId }).update(updatedProps);
        }));
    }
    return Q.allSettled(asyncCalls).then(x => {
        return Q.nbind(bulk.execute, bulk)().then(result => {
            // update parent
            return findMany(model, ids).then(objects => {
                return mongooseHelper.updateParent(model, objects).then(res => {
                    return objects;
                });
            });
        }).catch(error => {
            winstonLog_1.winstonLog.logError(`Error in bulkPut ${error}`);
            return Q.reject(error);
        });
    });
    //Enumerable.from(objArr).forEach(x => {
    //    if (x['_id']) {
    //        asyncCalls.push(put(model, x['_id'], x));
    //    }
    //});
    //return Q.allSettled(asyncCalls)
    //    .then(result => {
    //        return Enumerable.from(result).select(x => x.value).toArray();
    //    })
    //    .catch(error => {
    //        winstonLog.logError(`Error in bulkPut ${error}`);
    //        return Q.reject(error);
    //    });
}
exports.bulkPut = bulkPut;
/**
 * Iterate through objArr and call put for these
 * Usage - Update multiple object sequentially
 * @param model
 * @param objArr
 */
function bulkPatch(model, objArr) {
    var asyncCalls = [];
    var length = objArr.length;
    var ids = objArr.map(x => x._id);
    var bulk = model.collection.initializeUnorderedBulkOp();
    var asyncCalls = [];
    // classic for loop used gives high performanance
    for (var i = 0; i < length; i++) {
        asyncCalls.push(mongooseHelper.addChildModelToParent(model, objArr[i], objArr[i]._id).then(result => {
            var objectId = new Mongoose.Types.ObjectId(result._id);
            delete result._id;
            let clonedObj = mongooseHelper.removeTransientProperties(model, result);
            var updatedProps = Utils.getUpdatedProps(clonedObj, entity_change_1.EntityChange.patch);
            bulk.find({ _id: objectId }).update(updatedProps);
        }));
    }
    return Q.allSettled(asyncCalls).then(x => {
        return Q.nbind(bulk.execute, bulk)().then(result => {
            // update parent
            return findMany(model, ids).then(objects => {
                return mongooseHelper.updateParent(model, objects).then(res => {
                    return objects;
                });
            });
        }).catch(error => {
            winstonLog_1.winstonLog.logError(`Error in bulkPut ${error}`);
            return Q.reject(error);
        });
    });
}
exports.bulkPatch = bulkPatch;
/**
 * Bulk update objects. Find updated objects, and update the parent docs
 * Usage - Update multiple objects parallely with same porperty set
 * @param model
 * @param objIds
 * @param obj
 */
function bulkPutMany(model, objIds, obj) {
    delete obj._id;
    let clonedObj = mongooseHelper.removeTransientProperties(model, obj);
    // First update the any embedded property and then update the model
    var cond = {};
    cond['_id'] = {
        $in: objIds
    };
    var updatedProps = Utils.getUpdatedProps(clonedObj, entity_change_1.EntityChange.put);
    return Q.nbind(model.update, model)(cond, updatedProps, { multi: true })
        .then(result => {
        return findMany(model, objIds).then(objects => {
            return mongooseHelper.updateParent(model, objects).then(res => {
                return result;
            });
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in put ${error}`);
        return Q.reject(error);
    });
}
exports.bulkPutMany = bulkPutMany;
/**
 * Usage - Find all the objects
 * @param model
 */
function findAll(model) {
    return Q.nbind(model.find, model)({})
        .then(result => {
        return Utils.toObject(result);
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in findAll ${error}`);
        return Q.reject(error);
    });
}
exports.findAll = findAll;
/**
 * Query collection and then populate child objects with relationship
 * Usage - Search object with given condition
 * @param model
 * @param query
 */
function countWhere(model, query) {
    let queryObj = model.find(query).count();
    //winstonLog.logInfo(`findWhere query is ${query}`);
    return Q.nbind(queryObj.exec, queryObj)()
        .then(result => {
        // update embedded property, if any
        return Q.resolve(result);
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in countWhere ${error}`);
        return Q.reject(error);
    });
}
exports.countWhere = countWhere;
function distinctWhere(model, query) {
    let queryObj = model.find(query).distinct();
    //winstonLog.logInfo(`findWhere query is ${query}`);
    return Q.nbind(queryObj.exec, queryObj)()
        .then(result => {
        // update embedded property, if any
        return Q.resolve(result);
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in distinctWhere ${error}`);
        return Q.reject(error);
    });
}
exports.distinctWhere = distinctWhere;
/**
 * Query collection and then populate child objects with relationship
 * Usage - Search object with given condition
 * @param model
 * @param query
 * @param select {Array<string> | any} In case it is an array of string, it selects the specified keys mentioned in the array. In case  it is an Object, it sets the JS object as the projection key
 * @param sort
 * @param skip
 * @param limit
 */
function findWhere(model, query, select, queryOptions, toLoadChilds, sort, skip, limit) {
    var sel = {};
    if (select instanceof Array) {
        select.forEach(x => {
            sel[x] = 1;
        });
    }
    else if (select) {
        sel = select;
    }
    if (queryOptions) {
        if (queryOptions.limit != null)
            limit = queryOptions.limit;
        if (queryOptions.skip != null)
            skip = queryOptions.skip;
        if (queryOptions.sort != null)
            sort = queryOptions.sort;
    }
    let queryObj = model.find(query, sel);
    if (sort) {
        queryObj = queryObj.sort(sort);
    }
    if (skip) {
        queryObj = queryObj.skip(skip);
    }
    if (limit) {
        queryObj = queryObj.limit(limit);
    }
    //winstonLog.logInfo(`findWhere query is ${query}`);
    return Q.nbind(queryObj.exec, queryObj)()
        .then(result => {
        // update embedded property, if any
        if (toLoadChilds != undefined && toLoadChilds == false) {
            return Utils.toObject(result);
        }
        var asyncCalls = [];
        Enumerable.from(result).forEach(x => {
            asyncCalls.push(mongooseHelper.embeddedChildren(model, x, false));
        });
        return Q.allSettled(asyncCalls).then(r => {
            return Enumerable.from(r).select(x => Utils.toObject(x.value)).toArray();
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in findWhere ${error}`);
        return Q.reject(error);
    });
    // winstonLog.logInfo(`findWhere query is ${query}`);
    // return Q.nbind(model.find, model)(query)
    //     .then(result => {
    //         return toObject(result);
    //     }).catch(error => {
    //         winstonLog.logError(`Error in findWhere ${error}`);
    //         return error;
    //     });
}
exports.findWhere = findWhere;
/**
 * Usage - find object with given object id
 * @param model
 * @param id
 */
function findOne(model, id) {
    return Q.nbind(model.findOne, model)({ '_id': id })
        .then(result => {
        return mongooseHelper.embeddedChildren(model, result, false)
            .then(r => {
            return Utils.toObject(r);
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in findOne ${error}`);
        return Q.reject(error);
    });
}
exports.findOne = findOne;
/**
 * Usage - find the object with given {property : value}
 * @param model
 * @param fieldName
 * @param value
 */
function findByField(model, fieldName, value) {
    var param = {};
    param[fieldName] = value;
    return Q.nbind(model.findOne, model)(param)
        .then(result => {
        return mongooseHelper.embeddedChildren(model, result, false)
            .then(r => {
            return Utils.toObject(r);
        });
    }, err => {
        winstonLog_1.winstonLog.logError(`Error in findByField ${err}`);
        return Q.reject(err);
    });
}
exports.findByField = findByField;
/**
 * Usage - Find all the objects with given ids
 * @param model
 * @param ids
 */
function findMany(model, ids, toLoadEmbeddedChilds) {
    if (toLoadEmbeddedChilds == undefined) {
        toLoadEmbeddedChilds = false;
    }
    return Q.nbind(model.find, model)({
        '_id': {
            $in: ids
        }
    }).then((result) => {
        if (result.length !== ids.length) {
            var error = 'findmany - numbers of items found:' + result.length + 'number of items searched: ' + ids.length;
            winstonLog_1.winstonLog.logError(`Error in findMany ${error}`);
            return Q.reject(error);
        }
        var asyncCalls = [];
        if (toLoadEmbeddedChilds) {
            Enumerable.from(result).forEach(x => {
                asyncCalls.push(mongooseHelper.embeddedChildren(model, x, false));
            });
            return Q.allSettled(asyncCalls).then(r => {
                return Enumerable.from(r).select(x => Utils.toObject(x.value)).toArray();
            });
        }
        else {
            return Utils.toObject(result);
        }
    });
}
exports.findMany = findMany;
/**
 * find object with the given id.
 * Check if property has a relationship. If yes, then find all the child objects and return child
 * @param model
 * @param id
 * @param prop
 */
function findChild(model, id, prop) {
    return Q.nbind(model.findOne, model)({ '_id': id })
        .then(result => {
        var res = Utils.toObject(result)[prop];
        var metas = CoreUtils.getAllRelationsForTargetInternal(model_entity_1.getEntity(model.modelName));
        if (Enumerable.from(metas).any(x => x.propertyKey == prop)) {
            // create new object and add only that property for which we want to do eagerloading
            var result = {};
            result[prop] = res;
            return mongooseHelper.embeddedChildren(model, result, true)
                .then(r => {
                return result[prop];
            });
        }
        return res;
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in findChild ${error}`);
        return Q.reject(error);
    });
}
exports.findChild = findChild;
/**
 * case 1: all new - create main item and child separately and embed if true
 * case 2: some new, some update - create main item and update/create child accordingly and embed if true
 * @param obj
 */
function post(model, obj) {
    let clonedObj = mongooseHelper.removeTransientProperties(model, obj);
    return mongooseHelper.addChildModelToParent(model, clonedObj, null)
        .then(result => {
        try {
            mongooseHelper.autogenerateIdsForAutoFields(model, clonedObj);
        }
        catch (ex) {
            console.log(ex);
            return Q.reject(ex);
        }
        return Q.nbind(model.create, model)(new model(clonedObj)).then(result => {
            let resObj = Utils.toObject(result);
            Object.assign(obj, resObj);
            return obj;
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in post ${error}`);
        return Q.reject(error);
    });
}
exports.post = post;
/**
 * Delete object with given id. Check if, any children have deletecase=true, then delete children from master table
 * Update parent document for all the deleted objects
 * Usage - Delete object given id
 * @param model
 * @param id
 */
function del(model, id) {
    return Q.nbind(model.findOneAndRemove, model)({ '_id': id })
        .then((response) => {
        return mongooseHelper.deleteCascade(model, Utils.toObject(response)).then(x => {
            return mongooseHelper.updateEmbeddedOnEntityChange(model, entity_change_1.EntityChange.delete, response, null)
                .then(res => {
                return ({ delete: 'success' });
            });
        });
    })
        .catch(err => {
        winstonLog_1.winstonLog.logError(`delete failed ${err}`);
        return Q.reject({ delete: 'failed', error: err });
    });
}
exports.del = del;
/**
 * Sequetially delete the objects
 * @param modelte
 * @param ids
 */
function bulkDel(model, objs) {
    var asyncCalls = [];
    var ids = [];
    Enumerable.from(objs).forEach(x => {
        if (CoreUtils.isJSON(x)) {
            ids.push(x._id);
        }
        else {
            ids.push(x);
        }
    });
    ids.forEach(x => {
        asyncCalls.push(del(model, x));
    });
    return Q.allSettled(asyncCalls)
        .then(result => {
        var ret = [];
        Enumerable.from(result).forEach(x => {
            if (x.value)
                ret.push(x.value);
            if (x.reason)
                ret.push(x.reason);
        });
        return ret;
    })
        .catch(err => {
        winstonLog_1.winstonLog.logError(`bulkDel failed ${err}`);
        return Q.reject('bulkDel failed');
    });
}
exports.bulkDel = bulkDel;
/**
 * Check whether decorator is applied or not.
 * @param path
 * @param decorator
 * @param propertyKey
 */
function isDecoratorApplied(path, decorator, propertyKey) {
    var isDecoratorPresent = false;
    let repo = dynamic_repository_1.GetRepositoryForName(path);
    var repoEntity = repo && repo.getEntityType();
    var optimisticLock = repoEntity && utils_1.MetaUtils.getMetaData(repoEntity, decorator, propertyKey);
    if (optimisticLock) {
        isDecoratorPresent = true;
    }
    return isDecoratorPresent;
}
/**
 * Check if any child object need to be added, if yes, then add those child objects.
 * update the object with propertie. And then update the parent objects.
 * Usage - Update the object with given object id
 * @param model
 * @param id
 * @param obj
 */
function put(model, id, obj, path) {
    let clonedObj = mongooseHelper.removeTransientProperties(model, obj);
    // First update the any embedded property and then update the model
    return mongooseHelper.addChildModelToParent(model, clonedObj, id).then(result => {
        var updatedProps = Utils.getUpdatedProps(clonedObj, entity_change_1.EntityChange.put);
        let isDecoratorPresent = isDecoratorApplied(path, decorators_1.Decorators.OPTIMISTICLOCK, "put");
        let query = { '_id': id };
        if (isDecoratorPresent === true) {
            updatedProps["$set"] && delete updatedProps["$set"]["__v"];
            updatedProps["$inc"] = { '__v': 1 };
            query["__v"] = obj["__v"];
        }
        return Q.nbind(model.findOneAndUpdate, model)(query, updatedProps, { new: true })
            .then(result => {
            if (!result && isDecoratorPresent === true) {
                return Q.reject("You are trying to update with stale data,please try again after some time.");
            }
            return mongooseHelper.updateEmbeddedOnEntityChange(model, entity_change_1.EntityChange.put, result, Utils.getPropertiesFromObject(clonedObj))
                .then(res => {
                var resObj = Utils.toObject(result);
                return mongooseHelper.fetchEagerLoadingProperties(model, resObj).then(r => {
                    Object.assign(obj, r);
                    return obj;
                });
            });
        }).catch(error => {
            winstonLog_1.winstonLog.logError(`Error in put ${error}`);
            return Q.reject(error);
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in put ${error}`);
        return Q.reject(error);
    });
}
exports.put = put;
/**
 * Check if any child object need to be added, if yes, then add those child objects.
 * update the object with propertie. And then update the parent objects.
 * Usage - Update the object with given object id
 * @param model
 * @param id
 * @param obj
 */
function patch(model, id, obj, path) {
    let clonedObj = mongooseHelper.removeTransientProperties(model, obj);
    // First update the any embedded property and then update the model
    return mongooseHelper.addChildModelToParent(model, clonedObj, id).then(result => {
        var updatedProps = Utils.getUpdatedProps(clonedObj, entity_change_1.EntityChange.patch);
        let isDecoratorPresent = isDecoratorApplied(path, decorators_1.Decorators.OPTIMISTICLOCK, "patch");
        let query = { '_id': id };
        if (isDecoratorPresent === true) {
            updatedProps["$set"] && delete updatedProps["$set"]["__v"];
            updatedProps["$push"] && delete updatedProps["$push"]["__v"];
            updatedProps["$inc"] = { '__v': 1 };
            if (obj["__v"]) {
                query["__v"] = obj["__v"];
            }
        }
        return Q.nbind(model.findOneAndUpdate, model)(query, updatedProps, { new: true })
            .then(result => {
            if (!result && isDecoratorPresent === true) {
                return Q.reject("You are trying to update with stale data,please try again after some time.");
            }
            return mongooseHelper.updateEmbeddedOnEntityChange(model, entity_change_1.EntityChange.patch, result, Utils.getPropertiesFromObject(clonedObj))
                .then(res => {
                var resObj = Utils.toObject(result);
                return mongooseHelper.fetchEagerLoadingProperties(model, resObj).then(r => {
                    Object.assign(obj, r);
                    return obj;
                });
            });
        });
    }).catch(error => {
        winstonLog_1.winstonLog.logError(`Error in patch ${error}`);
        return Q.reject(error);
    });
}
exports.patch = patch;
